.github/workflows/tag-publish.yaml

```diff
@@ -1,10 +1,170 @@
+      LOCAL_HELM_REPO_PATH: /tmp/helm-repo
+      LOCAL_TERRAFORM_REGISTRY_PATH: /tmp/terraform-registry
+      ENVIRONMENT_NAME: ${{ inputs.environment || 'release-prod' }}
     steps:
-      - run: echo "Placeholder"
+      - name: Validate artifact tag
+        env:
+          SEMVER_REGEX: ^v?(?:0|[1-9]\d*)\.(?:0|[1-9]\d*)\.(?:0|[1-9]\d*)(?:-(?:(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+(?:[0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$
+        run: |
+          (echo "$ARTIFACT_TAG" | grep -qP "$SEMVER_REGEX") || (echo "The artifact tag $ARTIFACT_TAG is not a valid semver-coerced value"; exit 1)
+      - name: Validate environment
+        run: |
+          (echo "$ENVIRONMENT_NAME" | grep -qP '^publish-(prod|stage)$') || (echo "This workflow may only be ran from publishing environments"; exit 1)
+      - name: Checkout repo
+        uses: actions/checkout@v4
+        with:
+          ref: ${{ github.workflow_sha }} # Publishing tooling should all be the same version
+      - name: Setup tooling Go
+        uses: actions/setup-go@v5
+        with:
+          go-version-file: "./tooling/go.mod"
+          cache-dependency-path: "./tooling/go.sum"
+          check-latest: true
+      # Sync artifacts down (required by everything)
+      - name: Assume AWS role for uploading the artifacts
+        uses: aws-actions/configure-aws-credentials@010d0da01d0b5a38af31e9c3470dbfdabdecca3a # v4.0.1
+        with:
+          role-skip-session-tagging: true
+          aws-region: us-west-2
+          role-to-assume: ${{ vars.ARTIFACT_DOWNLOAD_AWS_ROLE }}
+          role-session-name: "artifact-download-${{ github.run_number }}"
+          role-duration-seconds: 900
+      - name: Download artifacts from S3
+        env:
+          ARTIFACT_BUCKET: ${{ vars.ARTIFACT_SOURCE_BUCKET }}
+        run: aws s3 cp "s3://$ARTIFACT_BUCKET/teleport-plugins/tag/$ARTIFACT_TAG/" "$LOCAL_ARTIFACTS_PATH" --recursive
+      # Binary artifact promotion
+      - name: Assume AWS role for uploading the artifacts
+        # This step is only supported in production as there is no staging version of Houston
+        if: ${{ env.ENVIRONMENT_NAME == 'publish-prod' }}
+        uses: aws-actions/configure-aws-credentials@010d0da01d0b5a38af31e9c3470dbfdabdecca3a # v4.0.1
+        with:
+          role-skip-session-tagging: true
+          aws-region: us-west-2
+          role-to-assume: ${{ vars.HOUSTON_UPLOAD_AWS_ROLE }}
+          role-session-name: "houston-upload-${{ github.run_number }}"
+          role-duration-seconds: 900
+      - name: Upload artifacts to Houston
+        # This step is only supported in production as there is no staging version of Houston
+        if: ${{ env.ENVIRONMENT_NAME == 'publish-prod' }}
+        env:
+          HOUSTON_BUCKET: ${{ vars.HOUSTON_BUCKET }}
+        run: |
+          aws s3 sync --acl public-read "$LOCAL_ARTIFACTS_PATH" s3://$HOUSTON_BUCKET/teleport-plugins/${ARTIFACT_TAG##*-v}/ \
```

Comment:
If you combine the role policies, you can sync directly between two S3 buckets. I've done this in the teleport.e houston publishing workflow: https://github.com/gravitational/teleport.e/blob/029dc542f83eaa7d0a2bb9b7c82b20d2fd71a4cd/.github/workflows/publish-to-houston.yaml#L36

Reply:
Won't both buckets still be in the same account whether it is the current houston setup (dronestorage and clientbuilds) or a new setup with teleport-prod and teleport-stage?

@wadells started calling the dronestorage bucket "pending" which I think is a good name - I know I've previously referred to it as a "staging" bucket, but that is not correct with the split we currently have with stage/prod. If we move on from the gravitational account, I expect we'd still have a pending bucket (in both prod and stage) and a publish bucket (in both prod and stage), so that copying from pending to publish would still be within a single account.

Reply:
> or a new setup with teleport-prod and teleport-stage

This is a completely new setup. The resources in the `gravitational` AWS account are slowly being moved into separate, more secure accounts. Now that we have a staging and prod environment it makes sense IMO to start using these new artifact storage/pending buckets exclusively where we have the workflows setup to do so. This is the first place that we're using them, but we'll eventually switch to them in teleport/teleport.e repo workflows as well.

> @wadells started calling the dronestorage bucket "pending" which I think is a good name

Yea that works for me. I'll update var names/references to "pending".

> I expect we'd still have a pending bucket (in both prod and stage) and a publish bucket (in both prod and stage)

We have a "pending" bucket for prod, and another for staging. This bucket is akin to "dronestorage.gravitational.io" in both structure and purpose. We don't really have a "publish" bucket in either account, or IMO a need for one.

We do have "clientbuilds.gravitational.io" but this is the backend storage for Houston rather than some kind of long term artifact storage for some internal purpose. It will (probably) eventually be deleted entirely after Houston is decommissioned and a customer grace period has elapsed.